backend/tester/src/test/scala/io/theorg/tester/OrgchartApiTestSuite.scala

```diff
@@ -0,0 +1,136 @@
+import zio.ZManaged
+import zio.Ref
+
+object OrgchartApiTestSuite extends DefaultRunnableSpec {
+
+  private val MaxNodesPerLevel = 3
+  private val AddNodeConcurrency = 4
+  private val MaxAddNodeRequests = 1000
+  private val targetGraphqlUrl = "http://localhost:8080/graphql"
+
+  final case class OrgChartNodeMapped(
+      id: UUID,
+      title: String,
+      parentId: Option[UUID]
+  )
+  private val apiUrl: Uri =
+    Uri.parse(targetGraphqlUrl).toOption.get
+
+  def withServer(
+      test: SttpBackend[Task, ZioStreams with WebSockets] => ZIO[
+        TestEnvironment,
+        Throwable,
+        TestResult
+      ]
+  ): ZIO[TestEnvironment, Throwable, TestResult] = {
+    val testJob =
+      for {
+        backend <- HttpClientZioBackend.managed()
+        response <- Managed.fromEffect(test(backend))
+      } yield response
+
+    testJob.useNow
+  }
+
+  val orgChartSelection = OrgChart.nodes(
+    (OrgChartNode.id.map(
+      UUID.fromString
+    ) ~ OrgChartNode.title ~ OrgChartNode.parentId.map(
+      (_.map(UUID.fromString))
+    )).mapN(OrgChartNodeMapped(_, _, _))
+  )
+
+  def getNodes(level: Int = 0) = Query.loadOrgChart(level)(
+    orgChartSelection
+  )
+
+  def addNode(title: String, parentId: Option[UUID]) = {
+    Mutation.addNode(AddNodeInput(title, parentId.map(_.toString())))(
+      AddNodeResult.orgChart(orgChartSelection)
+    )
+  }
+
+  private val calibanServerSpec =
+    suite("Caliban server Spec")(
+      testM("Empty on first run") {
+        withServer { backend =>
+          val response = getNodes().toRequest(apiUrl).send(backend)
+          assertM(response.map(_.code.code))(equalTo(200))
+          assertM(response.map(_.body).absolve.map(_.flatten.map(_.size)))(
+            equalTo(Some(0))
+          )
+        }
+      },
+      testM("Concurrent updates don't break the requirements") {
+        withServer { backend =>
+          for {
+            stateRef <- Ref.make(Set(Option.empty[UUID]))
+            _ <- ZIO.foreachParN_(AddNodeConcurrency)(0 to MaxAddNodeRequests)(
+              _ => {
+                stateRef.get
+                  .map(state => {
+                    val n = util.Random.nextInt(state.size)
+                    state.iterator.drop(n).next
+                  })
+                  .flatMap(nextParent =>
+                    addNode("node title", nextParent)
+                      .toRequest(apiUrl)
+                      .send(backend)
+                      .map(_.body)
+                      .absolve
+                      .map(_.flatten.flatten)
+                      .map(
+                        _.map(_.map(_.parentId))
+                          .getOrElse(List(Option.empty[UUID]))
+                          .distinct
+                          .toSet
+                      )
+                      .flatMap(latestState => stateRef.update(_ => latestState))
+                      .ignore
+                  )
+
+              }
```

Comment:
Consider extracting the body logic of the concurrent loop. 
Would it make sense to get rid of the `.map.map` and `.flatten.flatten` using monad transformers?

Reply:
It would if this was a production codebase we build on top of but this is just a quick setup that I don't plan to maintain